examples/CMakeLists.txt

```diff
@@ -1,25 +1,64 @@
 endif()
-add_subdirectory(adaptiveprecision-blockjacobi)
-add_subdirectory(ginkgo-overhead)
-add_subdirectory(ginkgo-ranges)
-add_subdirectory(ilu-preconditioned-solver)
-add_subdirectory(ir-ilu-preconditioned-solver)
-add_subdirectory(inverse-iteration)
-add_subdirectory(iterative-refinement)
-add_subdirectory(minimal-cuda-solver)
-add_subdirectory(mixed-precision-ir)
-add_subdirectory(nine-pt-stencil-solver)
-add_subdirectory(papi-logging)
-add_subdirectory(performance-debugging)
-add_subdirectory(poisson-solver)
-add_subdirectory(preconditioned-solver)
-add_subdirectory(preconditioner-export)
-add_subdirectory(simple-solver)
-add_subdirectory(simple-solver-logging)
-add_subdirectory(three-pt-stencil-solver)
+
+foreach(example ${EXAMPLES_LIST})
+    add_subdirectory(${example})
+endforeach()
+
+set(run_list "")
+if(GINKGO_RUN_EXAMPLES)
+    foreach(example ${EXAMPLES_LIST})
+        add_custom_target("run-${example}"
+            COMMAND ${CMAKE_CURRENT_BINARY_DIR}/${example}/${example} > ${CMAKE_CURRENT_BINARY_DIR}/${example}/${example}.out
+            WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/${example})
+    endforeach()
+    foreach(example ${EXAMPLES_LIST})
+        file(WRITE ${CMAKE_CURRENT_BINARY_DIR}/${example}/diff-command "#!/bin/bash
+diff <(sed -n '7,$p' ${CMAKE_CURRENT_BINARY_DIR}/${example}/${example}.out | sed -E 's/([0-9]+.)//g') <(sed -n '6,$p' ${CMAKE_SOURCE_DIR}/examples/${example}/doc/results.dox | head -n -4 | sed -E 's/([0-9]+.)//g')")
+        add_custom_target("diff-${example}"
+            COMMAND chmod +x ${CMAKE_CURRENT_BINARY_DIR}/${example}/diff-command && ${CMAKE_CURRENT_BINARY_DIR}/${example}/diff-command
+            WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/${example})
+    endforeach()
+    file(WRITE ${CMAKE_CURRENT_BINARY_DIR}/run_all_examples.sh
+        "NUM_EXAMPLES=$(wc -l < example-targets.txt)\n"
+        "for i in $(seq 1 $NUM_EXAMPLES)\n"
+        "do\n"
+        "  target_name=$(head -n $i example-targets.txt | tail -1)\n"
+        "  ex_target=run-\$\{target_name\}\n"
+        "  cmake --build ${CMAKE_BINARY_DIR} --target $ex_target\n"
+        "done\n")
+    add_custom_target("run_all_examples"
+        COMMAND chmod +x ${CMAKE_CURRENT_BINARY_DIR}/run_all_examples.sh && ${CMAKE_CURRENT_BINARY_DIR}/run_all_examples.sh
+        WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR})
+    file(WRITE ${CMAKE_CURRENT_BINARY_DIR}/validate_all_examples.sh
+        "NUM_EXAMPLES=$(wc -l < example-targets.txt)\n"
+        "for i in $(seq 1 $NUM_EXAMPLES)\n"
+        "do\n"
+        "  target_name=$(head -n $i example-targets.txt | tail -1)\n"
+        "  ex_target=diff-\$\{target_name\}\n"
+        "  cmake --build ${CMAKE_BINARY_DIR} --target $ex_target\n"
+        "done\n")
+    add_custom_target("validate_all_examples"
+        COMMAND chmod +x ${CMAKE_CURRENT_BINARY_DIR}/validate_all_examples.sh && ${CMAKE_CURRENT_BINARY_DIR}/validate_all_examples.sh
```

Comment:
I would prefer if we could have one validation target per example and let `validate_all_examples` depend on these individual targets - then we wouldn't need these intermediate bash scripts, and it would be easier to re-run just a single validation as a developer.

Reply:
Yes, each example has a separate validation target called `diff-<example_name>`. The `validate_all_examples.sh` just calls these targets. A dev can individually validate each example as well.

Reply:
Ah yes, I missed that. But we could still avoid the intermediate bash scripts by adding a dummy validate_all_examples target and using `add_dependency` to depend on all diff-* targets.